### ISSUE TITLE
Conventions

### ISSUE THREAD

**Issue by wwitzel** [PyProveIt/Prove-It#55]
Let's use this "issue" to have a dialog about conventions -- make suggestions, discuss alternatives, and come to agreements on best practices.

**Comment by wwitzel** [comment]
I propose the following convention for variable names based upon specified or intended type. 
Booleans: A, B, C, D, etc.
Complexes/Reals: a, b, c, d, etc.
Integers: i, j, k, l, m, n  (Should we always start with i?  I often start with k?  Is that all right?)

**Comment by wwitzel** [comment]
I've been using the following conventions for method naming:
If it derives a new truth given self as a truth: derive...
If it is used to conclude self as a truth: conclude...
If it deduces a truth but is neither of these: deduce...
There are exceptions.  Sometimes there is simply a better/cleaner name.  But if "derive", "deduce", or "conclude" is used, it should be consistent with this convention.

One of the exceptions is in the case of deducing equivalences or deriving something via an equivalence.  When deducing an equivalence, use nouns usually ending in "-ion".  Examples are:
substitution
commutation
association
factorization
distribution
When deriving something via an equivalence, use the verb form:
substitute
commute
associate
factorize
distribute
(You can call these on inner expressions of a known truth for example.  This is something I am currently working on and has not been pushed just yet).

**Comment by wwitzel** [comment]
Hoss noted the inconsistency of the following names of number set classes:
NaturalsSet, NaturalsPosSet, IntegerSet, RealsSet, RealsPosSet, RealsNegSet, and ComplexSet.
Should make it consistent with the following?
NaturalSet, NaturalPosSet, IntegerSet, RealSet, RealPosSet, RealNegSet, and ComplexSet.
And then their instantiations be (as is the current convention):
Naturals, NaturalsPos, Integers, Reals, RealsPos, RealsNeg, Complexes, etc.

Any preference?

**Comment by wdcraft01** [comment]
Seems OK … except for Complexes. Ick. I suppose ComplexNumbers is too long.
Just thinking out loud: Is there any use in considering NumberSet class names using SetOf… instead? Like SetOfNaturals, SetOfIntegers, SetOfPositiveIntegers, SetOfReals, … , SetOfComplexNumbers. Hmmm. I guess those are getting a little long.
Still don't like complexes, though.

________________________________

**Comment by wwitzel** [comment]
Complexes was Kenny Rudinger's decision a while ago.  It's kind of ugly, but it works after you get used to it.  I prefer that to "SetOfComplexNumbers".  Though I may be convinced that "domain=Complex" is better than "domain=Complexes".  I'd want to convince Kenny though -- he was pretty convinced that "Complexes" was the right way to go at the time.

**Comment by wdcraft01** [comment]
Permutations, Commutations, and such.
I did a little more review reading about permutations after our project meeting today, and it seems the kind of “pluck and insert” commutation or permutation we are producing in the `apply_commutation_thm` could be viewed as a cycle. For example, the permutation &sigma; defined by:
&sigma;({a, b, c, d, e}) = {a, d, b, c, e}
can be described in cycle notation as (b c d), or more explicitly as (a)(b c d)(e), where (a) is a 1-cycle, (b c d) is a 3-cycle, and (e) is another 1-cycle.
On a related note, it's difficult to find good examples of the noun “commutation” being used to describe the permuting of operands the way we're doing it. I do see references to a “commutator” and a “commutation operation,” which are quite different from what we have in mind.

**Comment by wwitzel** [comment]
Good.  But now there is a tense issue.  To follow the convention I've had, we need a noun to generate the equality/equivalence and we need a verb to perform the transformation of a known truth.  Since "cycle" is a noun and a verb, that is tricky.  Maybe we just use "permutation" and "permute" but have flexibility in the arguments that can be passed in.  Maybe something like

def permutation(new_order=None, cycle=None):
   ...

where you could do
permutation([1, 3, 5, 2, 0, 4])
to get 0,1,2,3,4,5 -> 1,3,5,2,0,4
or you could do
permutation(cycle=(1, 4))
to get 0,1,2,3,4,5 -> 0,2,3,4,1,5

If you provide both new_order and cycle, you should get an error.  If you provide neither, you should get an error.

I think I like this option.  What do you think?

**Comment by wwitzel** [comment]
Alternatively, we could have a separate method: "cyclicPermutation".  Maybe that's better.  I could go either way.

**Comment by wdcraft01** [comment]
My initial gut reaction to permutation(new_order=None, cycle=None) was that it seems redundant to offer both new_order and cycle possibilities, but after thinking about it a bit I like the idea that someone could use whichever notation they like or find convenient, in effect offering more flexibility for future interactions with other code.
In keeping with extant notation out in the world of permutations, the permutation(cycle=(1,4)) should (?) probably result in
`0, 1, 2, 3, 4, 5` &rarr; `0, 4, 2, 3, 1, 5`
(instead of the “pluck and insert” result of `0, 2, 3, 4, 1, 5`). To get the “pluck and insert” result, we would then need to use the new_order param, something like
`permutation(new_order=[0, 2, 3, 4, 1, 5])`
or the cycle param in this way:
`permutation(cycle=(1, 2, 3, 4))`
Something to consider as well would be how to effectively deal with multiple cycles (which are commutative), entertaining something like this:
`permutation(cycle=(0, 5)(1, 2, 3, 4))`
to produce:
`0, 1, 2, 3, 4, 5` &rarr; `5, 2, 3, 4, 1, 0`
Interesting to consider as well: our notation here in these examples confounds the indices with the elements or values. The cycle notation I believe traditionally uses values (and thus the cycle (0, 5) indicates the value 0 being mapped to the value 5 and vice versa) instead of locations. Just something to be clear about as we discuss.

**Comment by wwitzel** [comment]
I think we are on the right track and I trust you to make good decisions that properly respect conventions and make good sense.  I do feel strongly that we should use indices rather than values.  What do you do if a value is repeated??

**Comment by wdcraft01** [comment]
> I think we are on the right track and I trust you to make good decisions that properly respect conventions and make good sense. I do feel strongly that we should use indices rather than values. What do you do if a value is repeated??

Good point. Indices will avoid problems when dealing with repeated values.
On a related note, I was just realizing more fully that our Set class is really an ordered multi-set object, so we can create the set S1 = Set(a, b) = {a, b} and the set S2 = Set(a, b, a) = {a, b, a}. Manifested in that way, it seems clear that S1 &ncong; S2, and yet it's not absolutely clear how that would play out in Prove-It because set equivalence is defined as each set being an improper subset of the other, with improper subset being defined in the usual way (A &subseteq; B means that whenever a &in; A, then a &in; B). Clearly S1 &subseteq; S2. But probably S2 &subseteq; S1 as well, by our definition? And thus S1 &cong; S2 ?

**Comment by wwitzel** [comment]
What?  I would think {a, b} is set equivalent to {a, b, a}.  They are not strictly equal but they should be set equivalent.  Right??

**Comment by wwitzel** [comment]
BTW, when it comes to repeated values, think about added terms rather than sets.  For example,
a + b + a + c = b + a + a + c

**Comment by wdcraft01** [comment]
> What? I would think {a, b} is set equivalent to {a, b, a}. They are not strictly equal but they should be set equivalent. Right??

Yes, that makes sense. This is also then a more clear-cut example for me of the distinction between S1 = S2 versus S1 &cong; S2 (whereas in earlier discussions we often talked vaguely about the sets having some other properties that would make S1 ≠ S2 while S1 &cong; S2).

> BTW, when it comes to repeated values, think about added terms rather than sets. For example,
> a + b + a + c = b + a + a + c

That's a nice example to keep in mind. I have been too focused on just the set context.

**Comment by kmrudin** [comment]
Wayne is absolutely right- "Complexes" is my doing.  It is definitely a bit clunky, but if we're going to have "Reals" then I had thought "Complexes" is what correctly matches it.  Similarly, SetOfReals would match with SetOfComplexes, while SetOfRealNumbers would match with SetOfComplexNumbers (and RealNumbers would match with ComplexNumbers).  What I think is important is the ability for an end user to intuit what the domain name is, given other examples.  If we're using Integers and Reals, then I would have thought that Complexes would follow from that.  However, on further reflection this is perhaps not entirely analogous, as one typically refers to "the reals" or "the real numbers" alongside "the complex numbers" (but never "the complexes").  A cursory google search indicates that "complexes" is not widely used in the way I had wanted to use it.  (E.g., see Wolfram's entries for [complex number](http://mathworld.wolfram.com/ComplexNumber.html) and [reals](http://mathworld.wolfram.com/Reals.html).

I suppose we could go with "Integers", "Reals", "Floats", etc. and make the exception for "ComplexNumbers", as this accurately reflects how people actually talk.  There may not be a solution here that is a) internally consistent and b) not awkward/consistent with common parlance.

Another option is to always use the adjectival form of the word- "Real", "Complex", etc., but I worry that might introduce other issues (there may not always be non-clunky adjectival forms).

**Comment by wwitzel** [comment]
I've been thinking about whether Iter (for iterations) is a good name for what we are currently using it for:
x_1, ..., x_n
2^1, ..., 2^n
etc
'Iteration' has some broad connotations (iterative improvements, iterate until convergence, etc.).  I'm wondering if Seq (for sequence) would be more appropriate.  It conveys more of the consecutive nature of what we are dealing with.  My one hesitation with that is that the expanded form is also a "sequence", but our Seq would only be used in the "..." cases.  That is, "x_1, x_2, x_3" could be called a sequence, but it is not a Seq (it doesn't have the "...").  Anyway, I'd like your thoughts on this.

**Comment by wwitzel** [comment]
From kmrudin via e-mail:

I would avoid “Iter” as it has a python-specific meaning (https://docs.python.org/3/library/functions.html#iter).  

I think if there’s a “sequence” that isn’t a “Seq” this may cause confusion.  Is there a way to convey the *necessity* of the presence of ellipses in the name, or at least remind users that this is somehow different than a regular sequence?  What about something like “AbstractSeq” or “VarSeq” (as it’s a sequence of variable length)?

**Comment by wwitzel** [comment]
What about "Ellipses"??  It isn't too long and avoids ambiguity with other instances of showing a sequence.

**Comment by wwitzel** [comment]
kmrudin votes for EllipsisSeq.

**Comment by wwitzel** [comment]
Deepak's suggestion is "Series".  I like it!  It is an improvement over Iter, better conveying the mathematical sense of what it is, and it is short and rolls off the tongue.  We'll give it a few days to mull it over.  But I'm leaning toward adopting it right away.

**Comment by wwitzel** [comment]
Hoss strongly objected to "Series" because it typically implies a summation.  He has a good point, I'm afraid.  My latest suggestion is ExprRange.  Excerpt from my e-mail proposing this:

The "range" part tells you there is a start and an end.  It is consistent with Python's use of "range" being over integers (in the ExprRange case, the integers serve as "indices" going into a "Lambda" map).  Unfortunately, it is inconsistent with Python's "range" in that our "end" is inclusive and Python's is exclusive.  Well, nothing's perfect.   The "expr" part tells you that the elements are not anything specific -- they aren't necessarily numbers, or terms in a sum, or anything.

**Comment by wwitzel** [comment]
Another name-change proposal.  Deepak pointed out that our hypothetical reasoning rule is essentially the "deduction theorem".  I think the plan is to change HypotheticalReasoning to Deduction.

**Comment by wwitzel** [comment]
Another proposal.  I've been using 'aa', 'bb', 'cc', for bolded versions of 'a', 'b', and 'c', with the convention that the bolded versions are used for variable ranges.  However, I'm heading in the direction of always using explicit ranges like a_1, ..., a_n rather than the "a | |a|=n" style.  Using explicit ranges, there isn't such a need for bolded versions.  They are just more trouble than they are worth (it is easy to end up with errors caused by using 'a' instead of 'aa' for example).  Furthermore, if you want a bolded version, this should be done via style settings rather than different variables.

**Comment by wdcraft01** [comment]
> Another proposal. I've been using 'aa', 'bb', 'cc', for bolded versions of 'a', 'b', and 'c', with the convention that the bolded versions are used for variable ranges. ….

I would not miss the double-letter representation of bold (vector?) variables. I wonder, though, if it might still be good to have a tuple or vector or vector-like variable available. Do we have such a thing now, in addition to the iters? a_1,…,a_n could be quite clunky if we're talking about functions of vectors or vector outputs. Just a thought.

**Comment by wwitzel** [comment]
That's a good question.  I think the way to deal with proper linear algebra vectors is to make a kind of Vector operation and an maybe a indexing operation that is distinct from IndexedVar.  That is, although there are similarities between parameter "ranges" and a vector, there should be a distinction I think.  The "ranges" are treated at the core level.  A vector is an external linear algebra concept.  I'm not sure exactly how we want to flesh that out, but it should be distinct.  In the past, these have been conflated, using the R^n set notation with "ranges" formerly known as "iters" for example.  But let's think about building the vector concept as its own thing.
